x/mongo/driver/auth/oidc.go

```diff
@@ -0,0 +1,338 @@
+}
+
+func newOIDCAuthenticator(cred *Cred) (Authenticator, error) {
+	if cred.Password != "" {
+		return nil, fmt.Errorf("password cannot be specified for %q", MongoDBOIDC)
+	}
+	if cred.Props != nil {
+		if env, ok := cred.Props[environmentProp]; ok {
+			switch strings.ToLower(env) {
+			case azureEnvironmentValue:
+				fallthrough
+			case gcpEnvironmentValue:
+				if _, ok := cred.Props[resourceProp]; !ok {
+					return nil, fmt.Errorf("%q must be specified for %q %q", resourceProp, env, environmentProp)
+				}
+				fallthrough
+			case testEnvironmentValue:
+				if cred.OIDCMachineCallback != nil || cred.OIDCHumanCallback != nil {
+					return nil, fmt.Errorf("OIDC callbacks are not allowed for %q %q", env, environmentProp)
+				}
+			}
+		}
+	}
+	oa := &OIDCAuthenticator{
+		userName:                cred.Username,
+		AuthMechanismProperties: cred.Props,
+		OIDCMachineCallback:     cred.OIDCMachineCallback,
+		OIDCHumanCallback:       cred.OIDCHumanCallback,
+	}
+	return oa, nil
+}
+
+type oidcOneStep struct {
+	userName    string
+	accessToken string
+}
+
+func jwtStepRequest(accessToken string) []byte {
+	return bsoncore.NewDocumentBuilder().
+		AppendString("jwt", accessToken).
+		Build()
+}
+
+// TODO GODRIVER-3246: Implement OIDC human flow
+//func principalStepRequest(principal string) []byte {
+//	doc := bsoncore.NewDocumentBuilder()
+//	if principal != "" {
+//		doc.AppendString("n", principal)
+//	}
+//	return doc.Build()
+//}
+
+func (oos *oidcOneStep) Start() (string, []byte, error) {
+	return MongoDBOIDC, jwtStepRequest(oos.accessToken), nil
+}
+
+func (oos *oidcOneStep) Next([]byte) ([]byte, error) {
+	return nil, newAuthError("unexpected step in OIDC authentication", nil)
+}
+
+func (*oidcOneStep) Completed() bool {
+	return true
+}
+
+func (oa *OIDCAuthenticator) providerCallback() (OIDCCallback, error) {
+	env, ok := oa.AuthMechanismProperties[environmentProp]
+	if !ok {
+		return nil, nil
+	}
+
+	switch env {
+	// TODO GODRIVER-2728: Automatic token acquisition for Azure Identity Provider
+	// TODO GODRIVER-2806: Automatic token acquisition for GCP Identity Provider
+	// This is here just to pass the linter, it will be fixed in one of the above tickets.
+	case azureEnvironmentValue, gcpEnvironmentValue:
+		return func(ctx context.Context, args *OIDCArgs) (*OIDCCredential, error) {
+			return nil, fmt.Errorf("automatic token acquisition for %q not implemented yet", env)
+		}, fmt.Errorf("automatic token acquisition for %q not implemented yet", env)
+	}
+
+	return nil, fmt.Errorf("%q %q not supported for MONGODB-OIDC", environmentProp, env)
+}
+
+func (oa *OIDCAuthenticator) getAccessToken(
+	ctx context.Context,
+	conn driver.Connection,
+	args *OIDCArgs,
+	callback OIDCCallback,
+) (string, error) {
+	oa.mu.Lock()
+	defer oa.mu.Unlock()
+
+	if oa.accessToken != "" {
+		return oa.accessToken, nil
+	}
+
+	cred, err := callback(ctx, args)
+	if err != nil {
+		return "", err
+	}
+
+	oa.accessToken = cred.AccessToken
```

Comment:
There's a data race detected here when running the OIDC tests. We either need to use an `atomic.Value` to synchronize access to the access token value, or delegate all reading and writing of the access token to a function that synchronizes access to it.

See [data-race-2.txt](https://github.com/user-attachments/files/16057090/data-race-2.txt) for details.